src/style-spec/function/evaluation_context.js

```diff
@@ -97,13 +98,8 @@ module.exports = () => ({
 
     array: function(type, items) {
         if (!type) {
-            let itemtype;
-            for (const item of items) {
-                if (!itemtype) itemtype = typeof item;
-                else if (itemtype !== typeof item) itemtype = 'Value';
-            }
-            type = (!itemtype || itemtype === 'Value') ?
-                'Array' : `Array<${titlecase(itemtype)}>`;
+            const t = ArrayLiteral.inferArrayType(items);
```

Comment:
@jfirebaugh This change causes the runtime typing of `[get, foo]` when `foo` is an array to be as specific as possible, based on its actual contents:
 - the array is homogenous of type number, string, or boolean, then => `Array<typeof item, properties.foo.length>`
 - otherwise => `Array<Value, properties.foo.length>`